ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/HD/BackingStore.hs

```diff
@@ -0,0 +1,987 @@
+
+data Model values r = Model (Mock values) (HandleRefs r)
+  deriving stock Generic
+
+deriving instance (Show (Mock values), Show1 r) => Show (Model values r)
+
+initModel :: Operations keys values diff -> Model values r
+initModel ops = Model (emptyMock ops) []
+
+{------------------------------------------------------------------------------
+  Stepping the model
+------------------------------------------------------------------------------}
+
+transition ::
+     Eq1 r
+  => Operations keys values diff
+  -> Model values r
+  -> Cmd keys values diff :@ r
+  -> Resp keys values diff :@ r
+  -> Model values r
+transition ops m c = after . lockstep ops m c
+
+toMock :: (Functor f, Eq1 r) => Model values r -> f :@ r -> f MockHandle
+toMock (Model _ hs) (At fr) = (hs !) <$> fr
+
+step ::
+     Eq1 r
+  => Operations keys values diff
+  -> Model values r
+  -> Cmd keys values diff :@ r
+  -> (Resp keys values diff MockHandle, Mock values)
+step ops m@(Model mock _) c = runMock ops (toMock m c) mock
+
+data Event keys values diff r = Event {
+    before   :: Model values r
+  , cmd      :: Cmd keys values diff :@ r
+  , after    :: Model values r
+  , mockResp :: Resp keys values diff MockHandle
+  }
+
+lockstep ::
+     Eq1 r
+  => Operations keys values diff
+  -> Model values r
+  -> Cmd keys values diff :@ r
+  -> Resp keys values diff :@ r
+  -> Event keys values diff r
+lockstep ops m@(Model _ hs) c (At resp) = Event {
+      before = m
+    , cmd = c
+    , after = Model mock' (hs <> hs')
+    , mockResp = resp'
+    }
+  where
+    (resp', mock') = step ops m c
+    hs' = zip (toList resp) (toList resp')
+
+postcondition ::
+     (Show values, Eq values)
+  => Operations keys values diff
+  -> Model values Concrete
+  -> Cmd keys values diff :@ Concrete
+  -> Resp keys values diff :@ Concrete
+  -> Logic
+postcondition ops m c r = toMock (after e) r .== mockResp e
+  where
+    e = lockstep ops m c r
+
+{------------------------------------------------------------------------------
+  Constructing symbolic responses
+------------------------------------------------------------------------------}
+
+symbolicResp ::
+     Operations keys values diff
+  -> Model values Symbolic
+  -> Cmd keys values diff :@ Symbolic
+  -> GenSym (Resp keys values diff :@ Symbolic)
+symbolicResp ops m c = At <$> traverse (const genSym) resp
+  where
+    (resp, _mock') = step ops m c
+
+{------------------------------------------------------------------------------
+  Generating commands
+------------------------------------------------------------------------------}
+
+generator ::
+     forall keys values diff.
+     Operations keys values diff
+  -> Model values Symbolic
+  -> Maybe (Gen (Cmd keys values diff :@ Symbolic))
+generator ops model@(Model mock hs) = Just $ QC.oneof $
+      withoutHandle ++ withHandle
+  where
+    withoutHandle :: [Gen (Cmd keys values diff :@ Symbolic)]
+    withoutHandle = fmap At <$> [
+          pure BSClose
```

Comment:
Probably we should, though the precondition currently discards all generated `BSClose` commands. This is the case because backing stores will throw an uncaught error on operations like reads and writes once the backing store is closed, which makes tests fail. Furthermore, these errors are backingstore-specific error types, which means errors for each backing store implementation need to be handled separately.

A solution would be to catch these backingstore-specific errors in the `BackingStoreWithHandleRegistry` wrapper, and to return canonical ones, i.e., hypothetical errors like `LMDBBackingStoreClosedError` and `InMemBackingStoreClosedError` would both be turned into `BackingStoreClosedError`.

https://github.com/input-output-hk/ouroboros-network/pull/4081#discussion_r1000506172 is related, since some preconditions are also specific to backing store implementations.

To solve these problems, I will add more parameterisation to the property tests